### ISSUE TITLE
Handle new USER_DEACTIVATED and REALM_DEACTIVATED API error codes.

### ISSUE THREAD

**Issue by timabbott** [zulip/zulip-mobile#4571]
https://github.com/zulip/zulip/pull/17872 adds a couple of new custom error codes for errors that I'd expect us to see when trying to connect to a deactivated organization with the mobile apps (and also will change the error codes to 403).

We should check if these help the mobile app do better error handling for this error when trying to login or start the app.

**Comment by vaibhavs2** [comment]
Whats about if the implementation will be something like given below

| <div style="width:10px;">Description</div> | Layout|
| --- | --- | 
| If User logged In with realm and either realm or user account  is deactivated then app will open with this Account-Pick screen to either choose a valid account or add a new account | <img src="https://user-images.githubusercontent.com/31806059/113054025-938d5f80-91c6-11eb-8d8c-9c5e2d5e31b1.png" height="640" width="580"/> |
|While adding new realm if realm is deactivated | <img src="https://user-images.githubusercontent.com/31806059/113054631-43fb6380-91c7-11eb-9018-631d5ee45eda.png" height="640" width="580" /> |
<br/>
And same will be while authentication when user will try to login with user password then he would get Account deactivated error.

**Comment by chrisbobbe** [comment]
Yes, I think this should help us do better error handling when trying to login. In particular, this is our error handling when doing a password login:

```js
    try {
      const fetchedKey = await api.fetchApiKey({ realm, apiKey: '', email }, email, password);
      this.setState({ progress: false });
      dispatch(loginSuccess(realm, fetchedKey.email, fetchedKey.api_key));
    } catch (err) {
      this.setState({
        progress: false,
        error: requireEmailFormat
          ? 'Wrong email or password. Try again.'
          : 'Wrong username or password. Try again.',
      });
    }
```

I think we could helpfully inspect `err` to see if it's one of those new errors, before the existing `setState`, and if it's one of those, set the React state with an appropriate error message instead of what it's currently being set to.

Let's start with that, before looking at changes like those in @vaibhavs2's message. In particular,

- Remembering the deactivated state client-side (necessary for the first screenshot) is more than the new API is asking us to do. Could easily get complicated with requirements like doing the right thing when an account/realm is _re_activated, if that can happen
- For the second screenshot, I'm not sure we get `REALM_DEACTIVATED` on requesting `/server_settings`; my impression (from reading the linked PR a bit) is that the two errors are only given when you're trying to connect as a particular authenticated user, which we're not at that point in the UI.

**Comment by chrisbobbe** [comment]
I'd be happy to see a PR that made the logging-in improvement I described in my previous message. 🙂 

Also, though, @timabbott, my understanding is we'll receive these errors on any authenticated request, right, not just when logging in? So `POST /register`, etc.

If that's right, we can probably also be helpful by showing an alert alongside the code that forces a logout when we get a 4xx response. That's the two `dispatch(logout())` sites that aren't in `ProfileScreen` (the one for the user initiating a logout with a button).

**Comment by timabbott** [comment]
@chrisbobbe that is correct, these error codes appear when failing authentication with your API key, in addition to when you fail to get an API key due to deactivated accounts/realms.

**Comment by edgarcosta** [comment]
I should also add, that when sending messages to a group that includes a deactivated user one gets an infinite "sending" gif, but no reason is given why the message is not sent.

**Comment by gnprice** [comment]
> I should also add, that when sending messages to a group that includes a deactivated user one gets an infinite "sending" gif, but no reason is given why the message is not sent.

@edgarcosta Please file this as a separate issue, and then we'll be able to address it. Thanks!

**Comment by edgarcosta** [comment]
@gnprice , you know you could have just done that?

**Comment by gnprice** [comment]
@edgarcosta Posting a separate issue in an existing issue thread adds noise that gets in the way of resolving the original issue, and it doesn't help the new issue get fixed. In the future, please post new issues as their own new issue threads.

Please also review the Zulip project's [code of conduct](https://github.com/zulip/zulip/blob/main/CODE_OF_CONDUCT.md), and be mindful of maintaining a positive and respectful atmosphere in this community.
